content/client-lib-development-guide/features.textile

```diff
@@ -129,9 +129,12 @@ h3(#rest-auth). Auth
 * @(RSA5)@ TTL for new tokens is specified in milliseconds. If the user-provided @tokenParams@ does not specify a TTL, the TTL field should be omitted from the @tokenRequest@, and Ably will supply a token with a TTL of 60 minutes. See "TK2a":#TK2a
 * @(RSA6)@ The @capability@ for new tokens is JSON stringified. If If the user-provided @tokenParams@ does not specify capabilities, the @capability@ field should be omitted from the @tokenRequest@, and Ably will supply a token with the capabilities of the underlying key. See "TK2b":#TK2b
 * @(RSA7)@ @clientId@ and authenticated clients:
-** @(RSA7a)@ If a @clientId@ is provided in the @ClientOptions@, or is present in the current authentication token, then the client is considered to be identified (it has a @clientId@ that is implicit in all operations). Note that an authentication token @clientId@ wildcard value of @'*'@ is the exception where the client is unidentified and any @clientId@ is permitted in operations with Ably. The following applies to identified clients:
-*** @(RSA7a1)@ All operations (such as message publishing or presence) will have an implicit @clientId@. The Ably service automatically updates the @clientId@ attribute (when empty) for all @Message@ and @PresenceMessage@ messages received from that identified client, and any messages then published from the Ably service will have the @clientId@ attribute populated. It is therefore expected that Ably client libraries do not explicitly set the @clientId@ field on messages published when @clientId@ is implicit in the connection or authentication scheme
-*** @(RSA7a2)@ If @clientId@ is provided in @ClientOptions@, and an API key is provided along with no other means to generate a token, the client library will authenticate with Ably and obtain a token using the provided @clientId@ ensuring the token is restricted to operations for that @clientId@
+** @(RSA7d)@ If @clientId@ is provided in @ClientOptions@ and @RSA4@ indicates that token auth is to be used, the @clientId@ field in the @TokenParams@ (@TK2c@) should be set to that @clientId@ when requesting a token
+** @(RSA7e)@ If @clientId@ is provided in @ClientOptions@ and @RSA4@ indicates that basic auth is to be used, then:
+*** @(RSA7e1)@ For realtime clients, the connect request should include the @clientId@ as a querystring parameter, @clientId@
+*** @(RSA7e2)@ For REST clients, all requests should include an @X-Ably-ClientId@ header with value set to the @clientId@, Base64 encoded
```

Comment:
@SimonWoolf I assume you don't think it's necessary to specify how JS handles this?  Good suggestion re: Base64 btw.

Reply:
What's the reasoning behind having a header instead of a query param?

Reply:
as in, for when custom headers aren't supported (i.e. IE<10)? Nah, I don't think it's necessary to put it in the spec, it just complicates things for everyone else. Similar to how we do the `Authorization` header -- realtime also supports auth by qs param for old IEs, but the spec only mentions the main (header) way.

(I think this is justified because it's only temporary -- at some point, maybe not too far away, non-websocket webbrowsers will drop below ε% usage share, and we can drop all the jsonp, comet, upgrade etc. crap from ably-js entirely)

Reply:
> What's the reasoning behind having a header instead of a query param?

Consistency with the `Authorization: Basic ...` header (RSA11). Since that's how auth is already done for REST requests, adding a second header can be done just add it to the existing `Auth#getAuthHeaders` (or equivalent in another client lib) method.

Since there's currently no query param that should be applies to all rest requests, many client libs probably don't have an existing mechanism to add one, so doing it that way would be a bit more work.

(Though realtime will still have to support both for jsonp, so I'm trading off realtime ease of implementation for client lib ease of implementation)

Reply:
I too prefer headers, although I don't have a good reason why 😉 
It just feels a whole load cleaner that the REST URL endpoint is the same regardless of your identity.